### PR TITLE
Add support for media video and image parameters

### PR DESCRIPTION
**Pull request by keianhzo**
This PR adds support for missing media video parameters.

### CODE REVIEW

File: src/inflators/video-loader.ts
Comment (takahirox) [inline src/inflators/video-loader.ts:28]: Please use `!==` instead of `!=`.

File: src/bit-components.js
Comment (takahirox) [inline src/bit-components.js:177]: If I'm right, `projection` and `alphaMode` are not really needed to be strings in the component, aren't they? They seem to be used just as type (enum).

File: src/utils/load-audio.tsx
Comment (takahirox) [inline src/utils/load-audio.tsx:30]: I don't get these `MediaVideo.flags[eid]` and `MediaVideo.projection[eid]` checking are for. Does it mean an entity has `MediaVideo` component but thoese properties are not set? If so, how does it happen?

File: src/utils/load-image.tsx
Comment (takahirox) [inline src/utils/load-image.tsx:54]: I don't really get these `MediaImage` properties checking are for. An entity has `MediaImage` component but those properties are not set? If so, how does it happen?

Or, checking whether an entity has `MediaImage` component? If so, I don't think accessing component properties that is not assigned to an entity is safe. `hasComponent()` should be used instead.

File: src/inflators/video.ts
Comment (takahirox) [inline src/inflators/video.ts:40]: Thinking whether autoplay and loop property should be managed as flag in `MediaVideo` component because they are duplicated with `video.loop` and `video.autoplay`.

File: src/utils/jsx-entity.ts
Comment (takahirox) [inline src/utils/jsx-entity.ts:275]: Can't this line be `alphaMode: AlphaMode;`?

File: src/inflators/image-loader.ts
Comment (takahirox) [inline src/inflators/image-loader.ts:26]: An entity is assumed that it has `MediaImage` component? I don't think we should assume an entity in inflators alerady has certain components.

**Comment by takahirox** [comment]
I added "P1" (high priority) label to this PR.

I got a feedback from the team that the loop of an audio in a scene seems to have a problem that the audio would stop after a certain amount of time although it's expected to be looped.

This may be a very noticable problem with audio. Users will have bad user experience. I think this problem should be fixed by the initial release.

I guess this PR would fix the problem.

File: src/bit-components.js
Comment (keianhzo) [reply src/bit-components.js:177]: The enum is string based I think to make it easier to handle comparisons as those parameters as string in the glTFs.

File: src/utils/load-audio.tsx
Comment (keianhzo) [reply src/utils/load-audio.tsx:30]: Yeah this is very confusing. There are two path how we load images (same for the rest of media):
- Spawned Media:
**MediaPrefab -> inflateMediaLoader -> [media-loading-system] -> loadImage -> inflateImage (MediaImage component added)**
- glTF:
**inflateImageLoader -> inflateMediaLoader -> [media-loading-system] -> loadImage -> inflateImage (MediaImage component added)**

I am using `MediaImage` as a transport container for the `ImageLoaderParams` that are later gathered in `inflateImageLoader` and used in `loadImage` as it assumes that the `MediaImage` has been added an initialized already at that point. As you pointed out this is not correct as the `MediaImage` component has not been added yet.

I'm going to introduce a `MediaImageLoaderData` and `MediaVideoLoaderData` data stores in between `inflateImageLoader` and `loadImage` to safely pass the initialization parameters for all medias. I think that would solve most of the concerns in this PR.

File: src/utils/load-image.tsx
Comment (keianhzo) [reply src/utils/load-image.tsx:54]: See https://github.com/mozilla/hubs/pull/6111#discussion_r1288496997

File: src/inflators/image-loader.ts
Comment (keianhzo) [reply src/inflators/image-loader.ts:26]: https://github.com/mozilla/hubs/pull/6111#discussion_r1288496997

File: src/inflators/video.ts
Comment (keianhzo) [reply src/inflators/video.ts]: Yeah that makes sense, removed.

File: src/inflators/video.ts
Comment (takahirox) [inline src/inflators/video.ts:32]: Question: Is component data initialized as zero when the component is added to an entity? Eg: In case an entity is recycled. If not, we need `MediaVideo.flags[eid] = 0` above.

File: src/inflators/video.ts
Comment (takahirox) [inline src/inflators/video.ts:35]: Same above.

File: src/utils/load-audio.tsx
Comment (takahirox) [reply src/utils/load-audio.tsx:23]: Thanks for explaining it. Give me a little time to think whether we can simplify further...

File: src/bit-components.js
Comment (takahirox) [reply src/bit-components.js:177]: Can you elaborate this for me a bit more, please?

> handle comparisons as those parameters as string in the glTFs

What I though is, if I'm right, `projection` and `alphaMode` are finally used just for comparing like `projection === ProjectionMode.SPHERE_EQUIRECTANGULAR` or `switch(alphaMode) { case AlphaMode.Opaqueue: ...` in Hubs Client. So I don't think the enum values in Hubs Client really need to be strings. Instead they can be just integer and component data type can be simply `ui8`. Strings doesn't really fit to bitECS API and we use hack. Probably it would be good to avoid strings in component data where possible.

But this is kind of optimization. We may think of in another PR.

By the way, if you want to use strings, better to use `ui32` component data type. `ui8` can be easily shortage.

File: src/inflators/video.ts
Comment (takahirox) [inline src/inflators/video.ts:11]: Curious, any special reason why `1 << 2` here? For example, just `1` would look like good enough to me.

File: src/utils/jsx-entity.ts
Comment (takahirox) [inline src/utils/jsx-entity.ts:268]: Good catch

File: src/utils/load-audio.tsx
Comment (takahirox) [reply src/utils/load-audio.tsx:23]: Probably the root cause of the complexity is MediaLoader system is too big and operate too many things. If media type specific operation is needed, they can't be decoupled well from MediaLoader system.

For example, We may separate media loder system into media type specific loader system like ImageLoaderSystem, VideoLoaderSystem, ModelLoaderSystem, and common operations can be exported as utils that can be used from the new loaders. Each loader would be simplerer.

But I don't think we have time to refactor. So transporting as temporary solution would be ok.

The current `MediaImageLoaderData` and `MediaVideoLoaderData` approach may be a bit less maintainable. They are as if just global variables and future readers may feel hard to understand how they are properly managed. And it is decoupled from bitecs component and entity mechanism. So for example even if an component or entity is removed while the operation, the data is remained. It can be error prone, especially if the entity is recycled.

Perhaps managing the transporting data as component is safer? Give me more time for thinking proper temporary solution...

Comment (takahirox) [reply src/utils/load-audio.tsx:23]: Clean up idea https://github.com/mozilla/hubs/commit/3d2f38102ac052c1a7fa7834a02745d3b47bf750

* Treat MediaImage/VideoLoaderData as part of MediaLoader component data. Less problematic because their life cycle will be synched with MediaLoader component.
* Remove dependency with entity and MediaImage/VideoLoaderData from loaders in src/utils. The loaders will be simpler.

Feel free to cherry-pick if it looks good, or I can make a follow up PR after this PR lands.

File: src/inflators/video.ts
Comment (keianhzo) [reply src/inflators/video.ts:11]: Copy paste fail, fixed. Thanks.

File: src/inflators/video.ts
Comment (keianhzo) [reply src/inflators/video.ts:32]: Yes, it needs to be initialized, fixed. Thanks

File: src/inflators/video.ts
Comment (keianhzo) [reply src/inflators/video.ts:35]: Same. Fixed.

File: src/utils/load-audio.tsx
Comment (keianhzo) [reply src/utils/load-audio.tsx:23]: Thanks, sounds good to me. Maybe better to cleanup this after landing. Feel free to update that when this lands and I'll review.

File: src/bit-components.js
Comment (keianhzo) [reply src/bit-components.js]: Good call, I didn't pay much attention to it and I'm not sure why we decided to make them strings early on but I've migrated them to numbers to simplify components.

**Comment by keianhzo** [comment]
@takahirox I've addressed the feedback. When I have the final approval I'll land

File: src/utils/load-image.tsx
Comment (takahirox) [inline src/utils/load-image.tsx:53]: I guess misstyped here

```suggestion
    } else if ((params.alphaMode = ALPHA_MODE.MASK)) {
      imageDef.alphaMode === AlphaMode.Mask;
    }
```

File: src/utils/projection-mode.ts
Comment (takahirox) [inline src/utils/projection-mode.ts:5]: Maybe having `ProjectionMode` and `PROJECTION_MODE` enums may be confusing because the names are similar? How about renaming the latter to `ProjectionModeName` or something like that?

File: src/inflators/image.ts
Comment (takahirox) [inline src/inflators/image.ts:42]: Not a blocking issue but a comment for future clean up: Here can be similified to use a common pattern as [other inflators](https://github.com/mozilla/hubs/blob/master/src/inflators/ambient-light.ts) do

```typescript
export type XXXParams = {
  foo?: string;
  bar?: number;
};

const DEFAULTS: Required<XXXParams> = {
  foo: 1.0,
  bar: 'default'
};

function xxxSystem(world: HubsWorld, eid: EntityID, params: XXXParams) {
  const requiredParams = Object.assign({}, DEFAULTS, params) as Required<XXXParams>;
  addComponent(world, XXX, eid);
  XXX.foo[eid] = requiredParams.foo;
  XXX.bar[eid] = requiredParams.bar;
  ...
}
```

File: src/inflators/video-loader.ts
Comment (takahirox) [inline src/inflators/video-loader.ts:7]: Not blocking issue but just a question: Why only `projection` is optional in `VideoLoaderParams`?

File: src/utils/create-image-mesh.ts
Comment (takahirox) [inline src/utils/create-image-mesh.ts:16]: 1. Similar to `ProjectionMode`, maybe having `AlphaMode` and `ALPHA_MODE` enums may be confusing because the names are similar? How about renaming the latter to `AlphaModeName` or something like that?
2. Not a blocking issue but a future suggestion: There might be another properer place to define these enums, eg: `src/utils/alpha-mode.ts` similar to `src/utils/projection-mode.ts`.

File: src/inflators/image-loader.ts
Comment (takahirox) [inline src/inflators/image-loader.ts:9]: Nit: Maybe better to use `PROJECTION_MODE` and `ALPHA_MODE` types here? Probably you use strings type here because this inflator is (currently) only for glTF and no guarantee that valid data comes from glTF. But expected type will be clearer if we use `PROJECTION_MODE` and `ALPHA_MODE`. And it would be safer even if we end up using this inflator for jsx in the future.

SImilarly, in `video-loader.ts`.

**Review by takahirox — APPROVED** [review]
Added some minor comments but looking good. Please reflect the review comments (if needed) and merge. 

After this PR lands, I will make a follow up PR to clean up. https://github.com/mozilla/hubs/commit/7dc5aed3b5df12db4872af07f97806d3eab8fdd2

File: src/inflators/video-loader.ts
Comment (keianhzo) [reply src/inflators/video-loader.ts:7]: I don't think it is probably a copy&paste issue, updated

File: src/utils/create-image-mesh.ts
Comment (keianhzo) [reply src/utils/create-image-mesh.ts:16]: Updated

File: src/utils/load-audio.tsx
Comment (keianhzo) [reply src/utils/load-audio.tsx:21]: I think we only want FLAT projection for audio only video elements, right? It defaults to FLAT for audio.

**Comment by j-conrad** [comment]
Any chance of this getting deployed soon? Video and audio looping is becoming important for Behavior Graph testing as well as general Hubs scenes.

**Comment by keianhzo** [comment]
@j-conrad This is already deployed in the main Hubs client branch but not in the BGs branch. I'm waiting for us to land the new loader before rebasing and merging as there are still too often updates.

**Pull request merged**
